openquake/hazardlib/gsim/dowrickrhoades_2005.py

```diff
@@ -0,0 +1,318 @@
+    faulting types.
+    """
+
+    #: Supported tectonic region type for base class is 'active shallow crust'
+    DEFINED_FOR_TECTONIC_REGION_TYPE = const.TRT.ACTIVE_SHALLOW_CRUST
+
+    #: Supported intensity measure type is MMI.
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = set([MMI])
+
+    #: Supported intensity measure component is the horizontal component
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = const.IMC.HORIZONTAL
+
+    #: Supported standard deviation types are Inter, Intra and Total
+    # (see description in paragraph 5.2 page 200 )
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = set([
+        const.StdDev.TOTAL,
+        const.StdDev.INTER_EVENT,
+        const.StdDev.INTRA_EVENT
+    ])
+
+    #: The only site parameter is vs30 used to map to site class to distinguish
+    # between rock, stiff soil and soft soil
+    REQUIRES_SITES_PARAMETERS = set(('vs30', ))
+
+    #: Required rupture parameters are magnitude, and rake and hypocentral
+    # depth rake is for determining fault style flags. Hypo depth is for
+    # subduction GMPEs
+    REQUIRES_RUPTURE_PARAMETERS = set(('mag', 'rake', 'hypo_depth'))
+
+    #: Required distance measure is rrup (paragraphy x, page xx) which is
+    # defined as nearest distance to the source.
+    REQUIRES_DISTANCES = set(('rrup', ))
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """
+        See :meth:`superclass method
+        <.base.GroundShakingIntensityModel.get_mean_and_stddevs>`
+        for spec of input and result values.
+        """
+        assert all(stddev_type in self.DEFINED_FOR_STANDARD_DEVIATION_TYPES
+                   for stddev_type in stddev_types)
+
+        # extract dictionaries of coefficients specific to required
+        # intensity measure type
+        C = self.COEFFS[imt]
+
+        # Deltas for Tectonic Region Type and rake angles
+        delta_R, delta_S, delta_V, delta_I = self._get_deltas(rup.rake)
+
+        mean = self._compute_mean(C, rup.mag, dists.rrup, rup.hypo_depth,
+                                  delta_R, delta_S, delta_V, delta_I,
+                                  sites.vs30)
+
+        stddevs = self._get_stddevs(C, stddev_types, sites.vs30.size)
+
+        return mean, stddevs
+
+    def _compute_mean(self, C, mag, rrup, hypo_depth, delta_R, delta_S,
+                      delta_V, delta_I, vs30):
+
+        """
+        Compute MMI Intensity Value as per Equation in Table 5 and
+        Table 7 pag 198.
+        """
+        # mean is calculated for all the 4 classes using the same equation.
+        # For DowrickRhoades2005SSlab, the coefficients which don't appear in
+        # Model 3 equationare assigned to zero
+
+        mean = (C['A1'] + (C['A2'] + C['A2R'] * delta_R + C['A2V'] * delta_V) *
+                mag + (C['A3'] + C['A3S'] * delta_S + C['A3V'] * delta_V) *
+                np.log10(np.power((rrup**3 + C['d']**3), 1.0 / 3.0)) +
+                C['A4'] * hypo_depth + C['A5'] * delta_I)
```

Comment:
In general we have adopted a style of implementation that splits the GMPEs/IPEs into their separate components (e.g. magnitude term, distance term, style of faulting, site etc.). We do this now because we find it increasingly common for GMPEs to make small modifications on existing models or functional forms, so breaking the GMPE into separate functions for the particular terms makes it much easier and more efficient to modify them in sub-classes. So in this line we would prefer it to be split into two functions, e.g.:
```python
def _get_magnitude_scaling(self, C, mag, delta_R, delta_V):
    """
    Some comment (if appropriate)
    """
    return C["A1"] + (C['A2'] + C['A2R'] * delta_R + C['A2V'] * delta_V) * mag

def _get_distance_scaling(self, C, rrup, hypo_depth, delta_S, delta_V, delta_I):
    """
    Another comment (if appropriate)
    """
    return (C['A3'] + C['A3S'] * delta_S + C['A3V'] * delta_V) *\
        np.log10(np.power((rrup**3 + C['d']**3), 1.0 / 3.0)) +\
        C['A4'] * hypo_depth + C['A5'] * delta_I
```
Then you can just call the mean as:
```python
mean = (self._get_magnitude_term(C, mag, delta_R, delta_V) +
        self._get_distance_term(C, rrup, hypo_depth, delta_S, delta_V, delta_I)
```